pkg/providers/tinkerbell/reconciler/reconciler.go

```diff
@@ -173,62 +178,58 @@ func (r *Reconciler) DetectOperation(ctx context.Context, log logr.Logger, tinke
 	}
 
 	wantVersionChange := currentKCP.Spec.Version != tinkerbellScope.ControlPlane.KubeadmControlPlane.Spec.Version
-	cpWantScaleChange := *currentKCP.Spec.Replicas != *tinkerbellScope.ControlPlane.KubeadmControlPlane.Spec.Replicas
-	workerWantScaleChange, err := r.WorkerReplicasDiff(ctx, tinkerbellScope)
-	if err != nil {
-		return "", err
-	}
-	wantScaleChange := workerWantScaleChange || cpWantScaleChange
 
 	// The restriction that not allowing scaling and rolling is covered in webhook.
 	op := NoChange
-	switch {
-	case wantScaleChange:
-		op = ScaleOperation
-	case wantVersionChange:
+	if wantVersionChange {
 		op = K8sVersionUpgradeOperation
 	}
 	log.Info("Operation detected", "operation", op)
 	return op, nil
 }
 
-// WorkerReplicasDiff indicates if there's difference between current and desired worker node groups.
-func (r *Reconciler) WorkerReplicasDiff(ctx context.Context, tinkerbellScope *Scope) (bool, error) {
-	workerWantScaleChange := false
-	for _, wnc := range tinkerbellScope.ClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {
-		md := &clusterv1.MachineDeployment{}
-		mdName := clusterapi.MachineDeploymentName(tinkerbellScope.ClusterSpec.Cluster, wnc)
-		key := types.NamespacedName{Namespace: constants.EksaSystemNamespace, Name: mdName}
-		err := r.client.Get(ctx, key, md)
+func (r *Reconciler) omitTinkerbellMachineTemplates(ctx context.Context, tinkerbellScope *Scope) error { //nolint:gocyclo
+	currentKCP, err := controller.GetKubeadmControlPlane(ctx, r.client, tinkerbellScope.ClusterSpec.Cluster)
+	if err != nil {
+		return errors.Wrap(err, "failed to get kubeadmcontrolplane")
+	}
+
+	if currentKCP == nil || currentKCP.Spec.Version != tinkerbellScope.ControlPlane.KubeadmControlPlane.Spec.Version {
+		return nil
+	}
+
+	cpMachineTemplate, err := tinkerbell.GetMachineTemplate(ctx, clientutil.NewKubeClient(r.client), currentKCP.Spec.MachineTemplate.InfrastructureRef.Name, currentKCP.GetNamespace())
+	if err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrap(err, "failed to get controlplane machinetemplate")
+	}
+
+	if cpMachineTemplate != nil {
+		tinkerbellScope.ControlPlane.ControlPlaneMachineTemplate = nil
+		tinkerbellScope.ControlPlane.KubeadmControlPlane.Spec.MachineTemplate.InfrastructureRef.Name = cpMachineTemplate.GetName()
+	}
+
+	for i, wg := range tinkerbellScope.Workers.Groups {
+		machineDeployment, err := controller.GetMachineDeployment(ctx, r.client, tinkerbellScope.ClusterSpec.Cluster, wg.MachineDeployment.GetName())
 		if err != nil {
```

Comment:
Would this also be a scale update?

Reply:
It can be a scale up which is why we only check for nil/rolling case below on line 216. we omit the templates otherwise for scale up.